Iterating on the HIP description (#721)
hashgraph/hedera-improvement-proposal: HIP/hip-0000-add-software-version-to-event.md

---
hip: <HIP number (this is determined by the HIP editor)>
title: Add Software Version to Events
author: Edward Wertz <@edward-swirldslabs>
working-group: a list of the technical and business stakeholders' name(s) and/or username(s), or name(s) and email(s).
type: Standards Track
category: Core

> needs-council-approval: <Yes | No>

Reviewer: ```suggestion
needs-council-approval: Yes
```

status: Draft

> created: 4/19/2023

Reviewer: ```suggestion
created: 2023-04-19
```

discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/720
updated: <comma separated list of dates>

> requires: <HIP number(s)>
> replaces: <HIP number(s)>
> superseded-by: <HIP number(s)>

Reviewer: ```suggestion
```

---
hip: <HIP number (this is determined by the HIP editor)>
title: Add Software Version to Events
author: Edward Wertz <@edward-swirldslabs>
working-group: a list of the technical and business stakeholders' name(s) and/or username(s), or name(s) and email(s).
type: Standards Track
category: Core
needs-council-approval: <Yes | No>
status: Draft
created: 4/19/2023
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/720
updated: <comma separated list of dates>
requires: <HIP number(s)>
replaces: <HIP number(s)>
superseded-by: <HIP number(s)>
---

## Abstract

Add the software version to events that are serialized and to the event interface. This permits the application to
determine which version of software was used to create each event, and optionally to ignore events created by previous
versions.

## Motivation

> This change should come before the implementation of the preconsensus event stream. The preconsensus event stream
> will save events to disk prior to their reaching consensus. On restart of the application all events in the preconsensus
> event stream will be replayed to reach consensus. If the restart coincides with a software upgrade, the new software
> version will receive events created from the previous software version. This change would enable the application to
> decide whether to act on the transactions generated by the older software version.

Reviewer 2: ```suggestion
This change is a prerequisite for a planned platform feature called the preconsensus event stream. The preconsensus event stream writes events to disk prior to the events reaching consensus. When a node is
restarted, events from this stream will be replayed, starting with events that came after the state snapshot
loaded into the node, and ending with the last events that were saved prior to the node shutting down.

When the system performs an upgrade, this new system of replaying events will make it possible for events
created before the software upgrade to reach consensus after the software upgrade. (Currently, these events
are simply discarded). The change proposed by this HIP will enable the application to decide whether to act on the transactions generated by the older software version, or if those transactions should be ignored.
```

Author: Michael moved the hip to 721 which prevents sign off and commit of the suggestion.  Suggestion accepted and adopted in the next commit.

## Rationale

Currently, events which do not reach consensus prior to a software upgrade are lost and the transactions are never
handled. With the introduction of the preconsensus event stream, replayed events may reach consensus and the
transactions would become available for handling. Adding the version number to the event will allow the application
to decide how to handle the older transactions. Software upgrades can take several minutes and transactions submitted
prior to the software upgrade may have timed out. [1](https://github.com/hashgraph/hedera-services/issues/6026)

If this feature is not implemented, the application will not know when it is receiving events that have been generated

> by a different software version and would attempt to handle them in the same way.

Author: recommendation accepted and updated in the next commit.

## User stories

* John submits a transaction just as the mainnet undergoes a software upgrade. When John comes back after the
  software upgrade to check on the status of his transaction, he can see whether the transaction came to
  consensus and was handled. If the transaction failed to be handled, John can be given more specific reasons as to

>   why.

Reviewer 2: I would just write "This is an internal platform feature, and will not be something that users will directly interact with."

Author: I disagree.  I think the ability to have transactions come to consensus and be handled after an upgrade is user impacting.

Author: I've added your text before the user story.

## Specification

The binary serialization format for events will be updated to include the software version. The in-memory `Event`
interface will have a new method `getSoftwareVersion()` that returns the software version. This API change will affect
users that read our serialized event format or iterate over events in memory.

## Backwards Compatibility

> The data serialization change will be handled by the platform team. The addition of a new method on the Event interface
> will not impact backwards compatibility. The only expected impact would be to users that read the serialized data
> without using the platform software.

Reviewer 2: ```suggestion
Code previously used to deserialize the event stream (post consensus) will no longer be capable of parsing the new event stream without modification.
```

Author: updated text to reflect your suggestion, and added that the new software version will be able to read events from older software versions.

> ## Security Implications

Reviewer 2: I would just say "This change does not have security implications."

Author: Current text is like a reminder to not introduce bugs.   Kelly asked me to be detailed here.   @poulok, thoughts?

Currently, there is no software version in events and event processing code paths are not sensitive to the software
version.

At a data layer, adding the software version does not impact security. The software version upgrade which
introduces the software version in serialized events may have to handle versions of the older event serialization
format until all nodes have upgraded.

At the application layer, new code paths which key off the software version value can be created. The application

> developer will need to consider how to process events and transactions from different software versions and. The same

Author: ```suggestion
developer will need to consider how to process events and transactions from different software versions. The same
```

security practices against malicious nodes sending events would apply and the application developer would need to
ensure they are not introducing bugs or attack vectors if the application receives events from older or future
software versions.

## How to Teach This

> No additional teaching would be required. Javadoc will be provided on the new interface method. The primary impacted
> developers are the Hedera app developers.

Author: ```suggestion
Javadoc will be provided on the new interface method. The impacted developers are the Hedera app developers that use the platform's Event interface and anyone that consumes the serialized event data without using the platform software. 
```

## Reference Implementation

> Not created yet.

Reviewer 2: ```suggestion
This feature will be developed by the Swirlds Labs team.
```

Author: replaced text

## Rejected Ideas

> Throughout the discussion of a HIP, various ideas will be proposed which are not accepted. Those rejected ideas should
> be recorded along with the reasoning as to why they were rejected. This both helps record the thought process behind the
> final version of the HIP as well as preventing people from bringing up the same rejected idea again in subsequent
> discussions.
> 
> In a way, this section can be thought of as a breakout section of the Rationale section that focuses specifically on why
> certain ideas were not ultimately pursued.

Reviewer 2: ```suggestion
n/a
```

Author: updated text from template content.

## Open Issues

> While a HIP is in draft, ideas can come up which warrant further discussion. Those ideas should be recorded so people
> know that they are being thought about but do not have a concrete resolution. This helps make sure all issues required
> for the HIP to be ready for consideration are complete and reduces people duplicating prior discussions.

Reviewer 2: ```suggestion
n/a
```

Author: updated text from template content.
